### PR TITLE
Canvas having preview of type Visual

### PR DESCRIPTION
**Pull request by me-andre**
### Describe the background of your pull request

What does your pull request do? Does it solve a bug (which one?), add a feature?

### Additional context

Add any other context

### Governance 

- [ ] Documentation is added
- [ ] Test cases are added or updated

By submitting this pull request I confirm that:
- I've read the contributing document https://github.com/alkem-io/.github/blob/master/CONTRIBUTING.md.
- I've read and understand the Code of Conduct https://github.com/alkem-io/.github/blob/master/CODE_OF_CONDUCT.md.
- I understand that any contributions or suggestions I made may make it into the actual code. I've read the License 
     https://github.com/alkem-io/Coordination/blob/master/LICENSE.
- I understand that submitting the pull request requires agreeing to and signing the contributor license agreement.

### CODE REVIEW

File: src/migrations/1657204514370-canvas-visual.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+const table = '`canvas`';
+const column = '`bannerCardID`';
+
+export class canvasVisual1657204514370 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE ${table} ADD ${column} char(36) NOT NULL`
```
Comment (alstojanovic) [inline src/migrations/1657204514370-canvas-visual.ts]: You defined `bannerCard` as optional in entity and interface, adding NOT NULL in migration will basically make it mandatory replacing NULL values with empty strings. (actually not sure what char(36) null value would be)

File: src/domain/common/canvas/canvas.entity.ts [hunk 2]
```diff
@@ -33,4 +34,12 @@ export class Canvas extends NameableEntity implements ICanvas {
   })
   @JoinColumn()
   checkout?: CanvasCheckout;
+
+  @OneToOne(() => Visual, {
+    eager: true,
```
Comment (hero101) [inline src/domain/common/canvas/canvas.entity.ts:35]: don't do eager

File: src/domain/common/canvas/canvas.entity.ts [hunk 3]
```diff
@@ -33,4 +34,12 @@ export class Canvas extends NameableEntity implements ICanvas {
...
+    cascade: true,
+    onDelete: 'SET NULL',
+  })
+  @JoinColumn()
+  bannerCard?: Visual;
```
Comment (hero101) [inline src/domain/common/canvas/canvas.entity.ts]: why not just `visual`?

**Review by hero101 — CHANGES_REQUESTED** [review]

**Review by hero101 — DISMISSED** [review]
Eager suggests loading visuals all the time. I would suggest removing eager loading and adding a resolver instead.

File: src/domain/common/canvas/canvas.entity.ts (hunk 3 above)
Comment (me-andre) [reply src/domain/common/canvas/canvas.entity.ts]: It's the name @techsmyth  suggested.

File: src/domain/common/canvas/canvas.entity.ts (hunk 2 above)
Comment (me-andre) [reply src/domain/common/canvas/canvas.entity.ts:35]: Well, I just took the approach from the recent `TemplateInfo`.

File: src/migrations/1657204514370-canvas-visual.ts (hunk 1 above)
Comment (me-andre) [reply src/migrations/1657204514370-canvas-visual.ts]: I think the mistake is in the interface.

File: src/domain/common/canvas/canvas.entity.ts (hunk 2 above)
Comment (techsmyth) [reply src/domain/common/canvas/canvas.entity.ts:35]: In a pending PR the TemplateInfo is no longer eager loaded: see branch server-2035

File: src/domain/common/canvas/canvas.entity.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/common/canvas/canvas.entity.ts]: Correct, this banner is used on cards - so bannerCard is a far better name

Comment (hero101) [reply src/domain/common/canvas/canvas.entity.ts]: This name is coupling the API to this version of the front end

Comment (techsmyth) [reply src/domain/common/canvas/canvas.entity.ts]: bannerNarrow tells nothing, and I have repeatedly seen confusion related to this term. Suggestions for a better name?

Comment (hero101) [reply src/domain/common/canvas/canvas.entity.ts]: My first suggestion still stands.

File: src/domain/common/canvas/canvas.entity.ts (hunk 2 above)
Comment (me-andre) [reply src/domain/common/canvas/canvas.entity.ts:35]: Addressed.
@hero101 , @techsmyth , one question: do I understand correctly, that now for each N canvases in the list, we run N queries for a canvas with the visual?

File: src/domain/common/canvas/canvas.entity.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/common/canvas/canvas.entity.ts]: ok, then just visual for now...

Comment (me-andre) [reply src/domain/common/canvas/canvas.entity.ts]: To me, specifically to the Canvas, it's more of a `Preview`.

File: src/domain/common/canvas/canvas.entity.ts (hunk 2 above)
Comment (techsmyth) [reply src/domain/common/canvas/canvas.entity.ts:35]: A single query can be run from the client to get the URIs for all canvas templates. Does that answer your question?

Comment (hero101) [reply src/domain/common/canvas/canvas.entity.ts:35]: No, eager tells the orm to load this relation everytime the canvas has been loaded. It's not graphql related.

Comment (me-andre) [reply src/domain/common/canvas/canvas.entity.ts:35]: @techsmyth , @hero101 , I meant SQL queries. With the changes I made (removing eager), for a query of N canvases with bannerCards, does the backend run N+1 SQL queries? (1 for canvases, N for each canvas+visual).

Comment (hero101) [reply src/domain/common/canvas/canvas.entity.ts:35]: No, and yes.
No - the change you made does not produce this result.
Yes - naive graphql implementation does this and it's addressed with data loaders.

File: src/domain/common/canvas/canvas.entity.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/common/canvas/canvas.entity.ts]: Interesting point, actually preview makes sense as a name.

File: src/migrations/1657204514370-canvas-visual.ts (hunk 1 above)
Comment (me-andre) [reply src/migrations/1657204514370-canvas-visual.ts]: Made `bannerCard` optional.

File: src/domain/common/canvas/canvas.resolver.fields.ts [hunk 4]
```diff
@@ -16,4 +19,14 @@ export class CanvasResolverFields {
   async checkout(@Parent() canvas: ICanvas) {
     return await this.canvasService.getCanvasCheckout(canvas);
   }
+
+  @UseGuards(GraphqlGuard)
+  @ResolveField('bannerCard', () => IVisual, {
+    nullable: false,
```
Comment (alstojanovic) [inline src/domain/common/canvas/canvas.resolver.fields.ts]: This makes it mandatory in the graphql api. Have you tried querying without a bannerCard?

Comment (me-andre) [reply src/domain/common/canvas/canvas.resolver.fields.ts]: I see, this will confuse a potential consumer. Will change!

**Review by alstojanovic — APPROVED** [review]

**Pull request merged**